Mongo java driver upgrade

lbrdnk opened the pull request:
Closes #37713, #37209

### Description

With this PR merged, deployment using hidden members should work correctly. Problem was that old java driver, which was part of monger does not allow sessions to be used with hidden member deployments. Monger is now removed from dependencies.

Some namespaces were copied from monger however. Those should be either removed or further updated as part of follow-up work. Namespaces in question contains `TODO`s in docstrings.

### How to verify

Run the tests for `:mongo` driver. Hidden members deployment could be verified manually. Same holds for use of `directConnection` parameter

### Checklist

- [X] Tests have been added/updated to cover changes in this PR

lbrdnk committed:
tmp: patched monger for 4.11.1 mongo java driver

lbrdnk committed:
tmp: Update monger utils

Aggregation probably wont work now, but we are not using those from monger anyway. With this change in place I'm able to load needed namespaces and create test-data dataset successfully.

Commit contains lot of condo errors that should be resolved while porting.

lbrdnk committed:
WIP: Monger removed in favor of java driver wrapper

lbrdnk committed:
Update java driver wrapper

lbrdnk committed:
Update srv test

lbrdnk committed:
Update comments

lbrdnk committed:
Use non keywordized run-command for `dbms-version`

lbrdnk committed:
Fix according to e2e tests

lbrdnk committed:
Cleanup

lbrdnk commented on modules/drivers/mongo/test/metabase/driver/mongo/java_driver_wrapper_test.clj:
This file now contains tests from `mongo.util-test`.

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo/util.clj:
Removed in favor of `java_driver_wrapper.clj`.

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -1370,7 +1371,10 @@
+    ;; TODO: Fixme! In following expression we were previously creating BasicDBObject's. As part of Monger removal
+    ;;       in favor of plain mongo-java-driver we now create Documents. I believe following conversion was and is
+    ;;       responsible for #38181.
+    (mapv (fn [^org.bson.BsonValue v] (-> v .asDocument org.bson.Document.))
```
I believe this is a source of #38181.

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo/operators.clj:
This file is now only remnant of monger. It would make sense to remove it completely. It should be discussed first however. Specifically, if we are for the removal, do we want to use strings or keywords instead of symbols provided by this namespace. Based on the result, schemas in mongo.qp would have to be adjusted.

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo/java_driver_wrapper.clj:
```diff
@@ -0,0 +1,341 @@
+    (update-ssl-db-details db-details)))
+
+;; TODO: ADD explanation from notion doc.
+(defn db-details->connection-string
```
While WIP, more context on building connection string can be found in [this notion doc](https://www.notion.so/metabase/Mongo-without-monger-9f43f2c6a1e14be9bb49a82fd5d65eda?pvs=4). Its contents will be moved here, into docstring, later.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/java_driver_wrapper.clj:
```diff
@@ -0,0 +1,341 @@
+
+(defprotocol ConvertToDocument
+  (^org.bson.Document to-document [input] 
+    "Converts given piece of Clojure data to org.bson.Document usable by java driver."))
```
Are `String`s and `Boolean`s really `org.bson.Document`? That would be magic or worse.

lbrdnk committed:
Separate `java-driver-wrapper` into multiple namespaces

lbrdnk committed:
Fix semantic type inference for serialized json

lbrdnk committed:
Fix options for run-command

lbrdnk committed:
Cleanup json namespace

lbrdnk committed:
Cleanup conversion namespace

lbrdnk committed:
Cleanup operators

lbrdnk committed:
Update kondo in operators ns

lbrdnk committed:
Cleanup connection namespace

lbrdnk committed:
Cleanup mongo namespace

lbrdnk committed:
Cleanup util namespace

lbrdnk committed:
Add todo

lbrdnk committed:
Move session related code to util

lbrdnk committed:
Cleanup database namespace

lbrdnk committed:
Update docstring for conn string generation

lbrdnk committed:
Update docstrings

lbrdnk committed:
Update tests

lbrdnk committed:
Update linter for with macros

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/conversion.clj:
```diff
@@ -0,0 +1,139 @@
+
+  org.bson.Document
+  (from-document [input {:keys [keywordize] :or {keywordize true} :as opts}]
+    (reduce (if keywordize
+              (fn [m ^String k]
+                (assoc m (keyword k) (from-document (.get input k) opts)))
+              (fn [m ^String k]
+                (assoc m k (from-document (.get input k) opts))))
+            (ordered-map/ordered-map)
```
```suggestion
    (persistent! (reduce (if keywordize
                           (fn [m ^String k]
                             (assoc! m (keyword k) (from-document (.get input k) opts)))
                           (fn [m ^String k]
                             (assoc! m k (from-document (.get input k) opts))))
            (transient (ordered-map/ordered-map))
```

lbrdnk replied:
Thank you.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/conversion.clj:
```diff
@@ -0,0 +1,139 @@
+  nil
+  (to-document [_input] nil)
+
+  String
+  (to-document [input] input)
+
+  Boolean
+  (to-document [input] input)
+
+  java.util.Date
+  (to-document [input] input)
```
These look quite superfluous given that they are `Object`s and they don't change the behavior.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/database.clj:
```diff
@@ -0,0 +1,62 @@
+
+(set! *warn-on-reflection* true)
+
+(defn- fqdn?
+  "A very simple way to check if a hostname is fully-qualified:
+   Check if there are two or more periods in the name."
+  [host]
+  (<= 2 (-> host frequencies (get \. 0))))
```
Where is this logic from?

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/database.clj:
```diff
@@ -0,0 +1,62 @@
+          (:conn-uri database)            database            ; connection URI has all the parameters
+          (:conn-uri (:details database)) (:details database)
+          :else
+          (throw (Exception. (str "with-mongo-connection failed: bad connection details:"
+                                  (:details database)))))]
```
Why not `ex-info`?

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:
```diff
@@ -105,16 +107,16 @@
                              row
                              col-parts)]
-              (m.conversion/from-db-object val :keywordize)))
+              (mongo.conversion/from-document val true)))
```
Just a naked boolean looks suspicious, shouldn't it be an options map?

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:
```diff
@@ -146,8 +148,8 @@
   (let [coll      (.getCollection db coll)
-        pipe      (m.util/into-array-list (m.conversion/to-db-object stages))
-        aggregate (.aggregate coll session pipe BasicDBObject)]
+        pipe      (java.util.ArrayList. ^java.util.Collection (mongo.conversion/to-document stages))
```
I suggest  importing these classes to make this line shorter.

lbrdnk replied:
Done.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/execute.clj:
```diff
@@ -176,36 +178,19 @@
-    (.runCommand db kill-cmd)))
-
 (defn execute-reducible-query
   "Process and run a native MongoDB query."
```
I suggest documenting when things expect `mongo.connection/*mongo-client*` to be set.

lbrdnk replied:
Done.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:
```diff
@@ -1370,7 +1370,10 @@
   [^String s]
   (try
-    (mapv (fn [^org.bson.BsonValue v] (-> v .asDocument com.mongodb.BasicDBObject.))
+    ;; TODO: Fixme! In following expression we were previously creating BasicDBObject's. As part of Monger removal
+    ;;       in favor of plain mongo-java-driver we now create Documents. I believe following conversion was and is
+    ;;       responsible for #38181.
```
Can you make this more specific? Reading this, I'm not quite sure what is to be fixed (and how).

lbrdnk replied:
Done.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/util.clj:
```diff
@@ -1,262 +1,93 @@
+         skip (.skip skip)
+         batch-size (.batchSize (int batch-size))
+         sort-criteria (.sort (mongo.conversion/to-document sort-criteria)))
+       (mapv #(mongo.conversion/from-document % (select-keys opts [:keywordize])))))
```
Wouldn't the following work?

```suggestion
       (mapv #(mongo.conversion/from-document % opts))))
```

lbrdnk replied:
I wanted to separate `from-document` params from those of `do-find` to keep `from-document`s input "clean". That would make sense if there were using more parameters.

metamben commented on modules/drivers/mongo/src/metabase/driver/mongo/util.clj:
```diff
@@ -1,262 +1,93 @@
+(defn start-session!
+  "Start session on client `c`."
+  ^ClientSession [^MongoClient c]
+  (.. c startSession))
```
```suggestion
  (.startSession c))
```

lbrdnk replied:
Done.

metamben commented on modules/drivers/mongo/test/metabase/driver/mongo/connection_test.clj:
```diff
@@ -0,0 +1,143 @@
+    (is (false? (#'mongo.db/fqdn? "localhost")))
+    (is (false? (#'mongo.db/fqdn? "localhost.localdomain")))))
+
+;; TODO: remove this!
```
Why not just do it?

lbrdnk replied:
Thanks.

metamben commented on modules/drivers/mongo/test/metabase/driver/mongo/connection_test.clj:
```diff
@@ -0,0 +1,143 @@
+                    :authdb "authdb"
+                    :use-srv true}]
+    (testing "mongo+srv connection string used when :use-srv is thruthy"
+      (is (re-find #"\Qmongodb+srv://test-user:test-pass@test-host.place.com/datadb?authSource=authdb\E"
```
Isn't this what we usually write as `str/includes?`?

lbrdnk replied:
`re-find` is definitely not necessary here. Thanks.

metamben commented on modules/drivers/mongo/test/metabase/driver/mongo/connection_test.clj:
```diff
@@ -0,0 +1,143 @@
+                   (mongo.connection/db-details->connection-string db-details))))
+    (testing "Only fqdn may be used with mongo+srv"
+      (is (thrown-with-msg? Throwable
+                            #"\QUsing DNS SRV requires a FQDN for host\E"
```
```suggestion
                            #"Using DNS SRV requires a FQDN for host"
```

lbrdnk replied:
Done.

metamben commented on modules/drivers/mongo/test/metabase/driver/mongo/connection_test.clj:
```diff
@@ -0,0 +1,143 @@
+    (testing "Only fqdn may be used with mongo+srv"
+      (is (thrown-with-msg? Throwable
+                            #"\QUsing DNS SRV requires a FQDN for host\E"
+                            (-> db-details (assoc :host "localhost")
```
```suggestion
                            (-> db-details
                                (assoc :host "localhost")
```

metamben commented on modules/drivers/mongo/test/metabase/driver/mongo/conversion_test.clj:
```diff
@@ -0,0 +1,28 @@
+
+(set! *warn-on-reflection* true)
+
+(deftest transformation-test
```
```suggestion
(deftest ^:parallel transformation-test
```

metamben commented on modules/drivers/mongo/test/metabase/driver/mongo/database_test.clj:
```diff
@@ -0,0 +1,19 @@
+    (is (false? (#'mongo.db/fqdn? "localhost")))
+    (is (false? (#'mongo.db/fqdn? "localhost.localdomain")))))
+
+(deftest db-name-test
```
```suggestion
(deftest ^:parallel db-name-test
```

metamben commented on src/metabase/driver/util.clj:
```diff
@@ -596,15 +596,20 @@
+    ssl-context))
+
+(defn ssl-socket-factory
+  "Generates an `SocketFactory` with the custom certificates added"
```
Here, and in the previous docstring:
```suggestion
  "Generates a `SocketFactory` with the custom certificates added"
```

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo.clj:
```diff
@@ -53,18 +45,18 @@
-       ;; (this is required because (1) is true even if the database doesn't exist)
-       (contains? db-names (:db db-stats))))))
+  [_ db-details]
+  (mongo.connection/with-mongo-client [^MongoClient c db-details]
+    (let [db-names (mongo.util/list-database-names c)]
+      (mongo.connection/with-mongo-database [^MongoDatabase db db-details]
```
`MongoDatabase` class does not provide getter for `MongoClient` it was created with. `with-mongo-database` is meant as utility to avoid `(let [db (mongo.util/database client ...)` in calls of `with-mongo-client`, because we are using database without the need for client in various places.

As we need both in this function, I've modified the code to use the `let` form.

lbrdnk committed:
Update can-connect to use let form

- Avoid nested `with-mongo-database` call.
- Avoid creating a set used for searching for database name.

lbrdnk committed:
Remove redundant set use from describe-database

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo.clj:
```diff
@@ -200,65 +192,45 @@
-    {:tables (set (for [collection (disj (mdb/get-collection-names conn) "system.indexes")]
+  (mongo.connection/with-mongo-database [^MongoDatabase db database]
+    ;; TODO: factor next line!
+    {:tables (set (for [collection (disj (into #{} (mongo.util/list-collection-names db)) "system.indexes")]
```
I think `into` should rather be used when I do some transformation to the source collection. Thanks.

Here set operations are used just to remove one element from a set. Form in question was simplified, to use `(for [... :when...]` instead of set operations.

lbrdnk committed:
Change from-document keywordize default to false

lbrdnk committed:
Update modules/drivers/mongo/src/metabase/driver/mongo/connection.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

lbrdnk committed:
Remove log message translation

lbrdnk committed:
Update maybe-add-ssl-context-to-builder! to always return builder

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo/connection.clj:
```diff
@@ -0,0 +1,120 @@
+     "&connectTimeoutMS=" (driver.u/db-connection-timeout-ms)
+     "&serverSelectionTimeoutMS=" (driver.u/db-connection-timeout-ms)
+     (when ssl "&ssl=true")
+     (when (seq additional-options) (str "&" additional-options)))))
```
I've decided to create connection string first and then apply it to `MongoClientSettings$Builder` using its `.applyConnectionString` method in `mongo.connection/db-details->mongo-client-settings`. Then, if we are setting some options that are not exposed via connection string (currently only ssl context), those are set directly on builder.

I've resorted to this solution as I find it simpler than set everything directly through builders. With new driver version there are actually 6 different builder classes involved in creation of `MongoClientSettings` (`SslSettings$Builder` used for setting ssl context is just one of them).

[Docs for MongoClientSettings$Builder](http://mongodb.github.io/mongo-java-driver/4.11/apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.Builder.html)

lbrdnk committed:
Update modules/drivers/mongo/src/metabase/driver/mongo/connection.clj

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

lbrdnk committed:
Indent from-document

lbrdnk committed:
Use transient in from-document for building a map

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

lbrdnk committed:
Remove redundant ConvertToDocument extensions

lbrdnk committed:
Use oredered-map in do-find's sort

lbrdnk commented on modules/drivers/mongo/src/metabase/driver/mongo.clj:
```diff
@@ -200,65 +192,45 @@
+  (let [coll (mongo.util/collection db (:name table))]
+    (mongo.util/do-find coll {:limit metadata-queries/nested-field-sample-limit
+                             :skip 0
+                             :sort-criteria {:_id sort-direction}
```
I've used map instead of ordered map here because there is only one key set.

I've updated the `do-find` function handle either ordered map or sequence of kv pairs as argument for `:sort-criteria`. Also the docstring was upadted accordingly.

lbrdnk committed:
Use ex-info in details-normalized

lbrdnk committed:
Add imports and update comment in execute ns

lbrdnk committed:
Update fixme comment

lbrdnk committed:
Pass opts instead of a selection to from-document in do-find

lbrdnk committed:
Avoid unnecessary double dot call

lbrdnk committed:
Update connection test ns according to review remarks

lbrdnk committed:
Make tests parallel

lbrdnk committed:
Docstring update

lbrdnk committed:
Update docstring

lbrdnk committed:
Merge branch 'master' into mongo-java-driver-wrapper
